main.py

```diff
@@ -345,31 +339,87 @@ def writer(args, res_queue, input_file_path):
     out_base_path = os.path.join(args.exp_dir, prefix + "_" + args.result_fname)
     if args.task == "regression":
         channels = [0]
-        outfiles = [open(os.path.join(out_base_path + ".track.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".track.bedGraph")]
     elif args.task == "classification":
         channels = [1]
-        outfiles = [open(os.path.join(out_base_path + ".peaks.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".peaks.bedGraph")]
     elif args.task == "both":
         channels = [0, 1]
-        outfiles = [open(os.path.join(out_base_path + ".track.bedGraph"), "w"),
-                    open(os.path.join(out_base_path + ".peaks.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".track.bedGraph"),
+                    os.path.join(out_base_path + ".peaks.bedGraph")]
+
+    # Temp dir used to save temo files during multiprocessing.
+    temp_dir = tempfile.mkdtemp()
+    for channel in channels:
+        os.makedirs(os.path.join(temp_dir, str(channel)))
+
     while 1:
         if not res_queue.empty():
             item = res_queue.get()
             if item == 'done':
                 break
-            for channel in channels:
-                batch_bg = get_batch_bg(args, item, channel, intervals)
-                df_to_bedGraph(batch_bg, outfiles[channel])
+            keys, batch = item
+            batch_num = keys.numpy()[0]
+            #No multi processing
+            if args.num_workers == 0:
+                start = 0
+                end = batch.shape[0]
+                for channel in channels:
+                    batch_bg = save_to_bedgraph([start, end], item, channel, intervals, outfiles[channel], args.round)
+            else:
+                num_jobs = math.ceil(batch.shape[0] / args.batches_per_worker)
+                pool_size = 0
+
+                if args.num_workers == -1: # spawn pool of processes
+                    num_cpus = mp.cpu_count()
+                    pool_size = min(num_jobs, num_cpus)
+                else: # user specified # of processes
+                    pool_size = args.num_workers
+
+                pool = mp.Pool(pool_size)
+                tempfiles = [os.path.join(temp_dir, str(channel), "{0:05}".format(num+batch_num)) for channel in channels for num in range(num_jobs)]
+                tmp_batch_ranges = [[i*args.batches_per_worker, (i+1)*args.batches_per_worker] for i in range(num_jobs)]
+                tmp_batch_ranges[-1][1] = batch.shape[0]
+                tmp_batch_ranges = tmp_batch_ranges*len(channels)
+               
+                all_channels = [channel for channel in channels for i in range(num_jobs)]
+                map_args = list(zip(tmp_batch_ranges, [item]*len(tmp_batch_ranges), all_channels, [intervals]*len(tmp_batch_ranges),
```

Comment:
instead of doubling all the ranges by number of channels, it might be more readable if we call map_args once with channel = 0, and another time with channel = 1. This way it becomes a bit more readable and avoids copying of the intervals every loop

Reply:
done.